### PR TITLE
CLOUDP-118714 register flow - regenerate auth code when expired

### PR DESCRIPTION
**Pull request by tibulca**
## Proposed changes

_Jira ticket:_ CLOUDP-118714

Closes #[issue number]

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongocli/blob/master/CONTRIBUTING.md) (if appropriate)
- [x] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [x] I have updated [e2e/E2E-TESTS.md](https://github.com/mongodb/mongocli/blob/master/e2e/E2E-TESTS.md) (if an e2e test has been added)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: internal/cli/auth/register.go [hunk 1]
```diff
@@ -20,52 +20,85 @@ import (
...
 		}
+
+		accessToken, _, err := opts.login.flow.PollToken(ctx, code)
+		if retry, errRetry := opts.shouldRetryRegister(err); errRetry != nil {
+			return errRetry
+		} else if retry {
+			continue
+		}
+
+		if err != nil {
```
Comment (blva) [inline internal/cli/auth/register.go:84]: [q] should this be on line 78?

**Review by blva — APPROVED** [review]
LGTM - 1 minor comment

File: internal/cli/auth/register.go (hunk 1 above)
Comment (tibulca) [reply internal/cli/auth/register.go:84]: at line 78 we check if `err` is a retriable error and if it is we check if we were able to read the user's answer (`errRetry `).
Here (line 84) we deal with the case when a non-retriable error occurred (e.g. cannot access the api anymore) or the user said he doesn't want to retry anymore (in which case we display the last error: token expired)

Comment (blva) [reply internal/cli/auth/register.go:84]: Thanks, I missed that we're passing error to the retry check.

File: internal/cli/auth/login.go [hunk 2]
```diff
@@ -172,7 +172,8 @@ func (opts *LoginOpts) oauthFlow(ctx context.Context) error {
 
 	accessToken, _, err := opts.flow.PollToken(ctx, code)
 	var target *atlas.ErrorResponse
-	if errors.As(err, &target) && target.ErrorCode == authExpiredError {
+	tokenExpired := err == auth.ErrTimeout || (errors.As(err, &target) && target.ErrorCode == authExpiredError)
```
Comment (gssbzn) [inline internal/cli/auth/login.go:175]: [non-blocking] feel free to add a function to https://github.com/mongodb/go-client-mongodb-atlas/tree/master/auth, `auth.IsTimeoutErr` or something

File: internal/cli/auth/register.go [hunk 3]
```diff
@@ -20,52 +20,85 @@ import (
...
+	"go.mongodb.org/atlas/auth"
 	atlas "go.mongodb.org/atlas/mongodbatlas"
 )
 
 const accountURI = "https://account.mongodb.com/account/register?fromURI=https://account.mongodb.com/account/connect"
 const govAccountURI = "https://account.mongodbgov.com/account/register?fromURI=https://account.mongodbgov.com/account/connect"
 
+type registerSurvey struct {
+	confirm func(message string, defaultResponse bool) (response bool, err error)
+}
```
Comment (gssbzn) [inline internal/cli/auth/register.go:37]: since the point here is to mock I'd use an interface rather than using functions but not a major blocker

**Review by gssbzn — APPROVED** [review]
LGTM

**Review by andreaangiolillo — APPROVED** [review]
LGTM

File: internal/cli/auth/login.go (hunk 2 above)
Comment (tibulca) [reply internal/cli/auth/login.go:175]: will do in a separate PR, thx

**Pull request merged**